8274242: Implement fast-path for ASCII-compatible CharsetEncoders on x86

cl4es opened the pull request:
This patch extends the `ISO_8859_1.implEncodeISOArray` intrinsic on x86 to work also for ASCII encoding, which makes for example the `UTF_8$Encoder` perform on par with (or outperform) similarly getting charset encoded bytes from a String. The former took a small performance hit in JDK 9, and the latter improved greatly in the same release.

Extending the `EncodeIsoArray` intrinsics on other platforms should be possible, but I'm unfamiliar with the macro assembler in general and unlike the x86 intrinsic they don't use a simple vectorized mask to implement the latin-1 check. For example aarch64 seem to filter out the low bytes and then check if there's any bits set in the high bytes. Clever, but very different to the 0xFF80 2-byte mask that an ASCII test wants. 

<!--
Replace this text with a description of your pull request (also remove the surrounding HTML comment markers).
If in doubt, feel free to delete everything in this edit box first, the bot will restore the progress section as needed.
-->

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8274242](https://bugs.openjdk.java.net/browse/JDK-8274242): Implement fast-path for ASCII-compatible CharsetEncoders on x86


### Reviewers
 * [Naoto Sato](https://openjdk.java.net/census#naoto) (@naotoj - **Reviewer**) ⚠️ Review applies to 8edc228f64fda449b7e6c65858241d7c3571bd4d
 * [Tobias Hartmann](https://openjdk.java.net/census#thartmann) (@TobiHartmann - **Reviewer**) ⚠️ Review applies to 4ad9c08ac90bdd2518ea111bd0cb1feeb8e0cc07


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/5621/head:pull/5621` \
`$ git checkout pull/5621`

Update a local copy of the PR: \
`$ git checkout pull/5621` \
`$ git pull https://git.openjdk.java.net/jdk pull/5621/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 5621`

View PR using the GUI difftool: \
`$ git pr show -t 5621`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/5621.diff">https://git.openjdk.java.net/jdk/pull/5621.diff</a>

</details>

wangweij commented:
/label remove security

cl4es commented:
The current version (cef05f4) copies the `ISO_8859_1.implEncodeISOArray` intrinsic and adapts it to work on ASCII encoding, which makes the `UTF_8$Encoder` perform on par with (or outperform) encoding from a `String`. Using microbenchmarks provided by @carterkozak  here: https://github.com/carterkozak/stringbuilder-encoding-performance

Baseline:
```
Benchmark                                                      (charsetName)                          (message)  (timesToAppend)  Mode  Cnt     Score    Error  Units
EncoderBenchmarks.charsetEncoder                                       UTF-8     This is a simple ASCII message                3  avgt    8   270.237 ± 10.504  ns/op
EncoderBenchmarks.charsetEncoder                                       UTF-8  This is a message with unicode 😊                3  avgt    8   568.353 ±  2.331  ns/op
EncoderBenchmarks.charsetEncoderWithAllocation                         UTF-8     This is a simple ASCII message                3  avgt    8   324.889 ± 17.466  ns/op
EncoderBenchmarks.charsetEncoderWithAllocation                         UTF-8  This is a message with unicode 😊                3  avgt    8   633.720 ± 22.703  ns/op
EncoderBenchmarks.charsetEncoderWithAllocationWrappingBuilder          UTF-8     This is a simple ASCII message                3  avgt    8  1132.436 ± 30.661  ns/op
EncoderBenchmarks.charsetEncoderWithAllocationWrappingBuilder          UTF-8  This is a message with unicode 😊                3  avgt    8  1379.207 ± 66.982  ns/op
EncoderBenchmarks.toStringGetBytes                                     UTF-8     This is a simple ASCII message                3  avgt    8    91.253 ±  3.848  ns/op
EncoderBenchmarks.toStringGetBytes                                     UTF-8  This is a message with unicode 😊                3  avgt    8   519.489 ± 12.516  ns/op
```

Patch:
```
Benchmark                                                      (charsetName)                          (message)  (timesToAppend)  Mode  Cnt     Score     Error  Units
EncoderBenchmarks.charsetEncoder                                       UTF-8     This is a simple ASCII message                3  avgt    4    82.535 ±  20.310  ns/op
EncoderBenchmarks.charsetEncoder                                       UTF-8  This is a message with unicode 😊                3  avgt    4   522.679 ±  13.456  ns/op
EncoderBenchmarks.charsetEncoderWithAllocation                         UTF-8     This is a simple ASCII message                3  avgt    4   127.831 ±  32.612  ns/op
EncoderBenchmarks.charsetEncoderWithAllocation                         UTF-8  This is a message with unicode 😊                3  avgt    4   549.343 ±  59.899  ns/op
EncoderBenchmarks.charsetEncoderWithAllocationWrappingBuilder          UTF-8     This is a simple ASCII message                3  avgt    4  1182.835 ± 153.735  ns/op
EncoderBenchmarks.charsetEncoderWithAllocationWrappingBuilder          UTF-8  This is a message with unicode 😊                3  avgt    4  1416.407 ± 130.551  ns/op
EncoderBenchmarks.toStringGetBytes                                     UTF-8     This is a simple ASCII message                3  avgt    4    97.770 ±  15.742  ns/op
EncoderBenchmarks.toStringGetBytes                                     UTF-8  This is a message with unicode 😊                3  avgt    4   516.351 ±  58.580  ns/op
```

This can probably be simplified further, say by adding a flag to the intrinsic of whether we're encoding ASCII only or ISO-8859-1. It also needs to be implemented and tested on all architectures.

cl4es commented:
> This can probably be simplified further, say by adding a flag to the intrinsic of whether we're encoding ASCII only or ISO-8859-1.

Done: Removed the addition of a new C2 Node, extended the existing macro assembler function to alter the mask.

> It also needs to be implemented and tested on all architectures.

Implementing this on other hardware is Future Work. The non-x86 intrinsics for implEncodeISOArray all seem to use clever tricks rather than a simple mask that can be easily switched from detecting non-latin-1(0xFF00) to detecting ASCII (0xFF80). Clever tricks make it rather challenging to extend this like I could easily do in the x86 code (most all assembler is foreign to me)

cl4es commented:
/label add hotspot-compiler

naotoj reviewed (approved):
core library part of the changes looks good to me.

sviswa7 commented:
x86 part of changes look good.

TobiHartmann commented on src/hotspot/cpu/x86/x86_32.ad:
```diff
@@ -12199,18 +12199,35 @@ instruct string_inflate_evex(Universe dummy, eSIRegP src, eDIRegP dst, eDXRegI l
+// encode char[] to byte[] in ASCII
+instruct encode_ascii_array(eSIRegP src, eDIRegP dst, eDXRegI len,
+                          regD tmp1, regD tmp2, regD tmp3, regD tmp4,
+                          eCXRegI tmp5, eAXRegI result, eFlagsReg cr) %{
```
Indentation is wrong.

TobiHartmann commented on src/hotspot/cpu/x86/x86_32.ad:
```diff
@@ -12199,18 +12199,35 @@ instruct string_inflate_evex(Universe dummy, eSIRegP src, eDIRegP dst, eDXRegI l
+  match(Set result (EncodeISOArray src (Binary dst len)));
+  effect(TEMP tmp1, TEMP tmp2, TEMP tmp3, TEMP tmp4, USE_KILL src, USE_KILL dst, USE_KILL len, KILL tmp5, KILL cr);
+
+  format %{ "Encode array $src,$dst,$len -> $result    // KILL ECX, EDX, $tmp1, $tmp2, $tmp3, $tmp4, ESI, EDI " %}
```
You might want to change the opto assembly comment to "Encode ascii array" (and to "Encode iso array" above). Same on 64-bit.

TobiHartmann commented on src/hotspot/share/opto/intrinsicnode.hpp:
```diff
@@ -170,8 +170,12 @@ class HasNegativesNode: public StrIntrinsicNode {
 //------------------------------EncodeISOArray--------------------------------
 // encode char[] to byte[] in ISO_8859_1
```
Comment should be adjusted to `... in ISO_8859_1 or ASCII`.

TobiHartmann reviewed (approved):
Very nice. The changes look good to me, just added some minor comments.

Should we remove the "iso" part from the method/class names?

simonis commented on src/hotspot/share/opto/c2compiler.cpp:
```diff
@@ -216,6 +216,10 @@ bool C2Compiler::is_intrinsic_supported(const methodHandle& method, bool is_virt
+  case vmIntrinsics::_encodeAsciiArray:
+#if !defined(X86)
+    return false; // not yet implemented
+#endif
```
It might be a little more work, but I think it's cleaner to move the decision whether the intrinisc is supported into the Matcher like for most other intrinsics and keep this code here platform independent. Otherwise we will get an increasing cascade of ifdefs as people start implementing this for other platforms.

cl4es replied:
Not too much work. I recently introduced platform-specific `matcher_*.hpp` files then adding a boolean constant is easy (no need to much with the .ad files) and more or less free since such constants are no longer defined in a generated .cpp file which disables inlining and folding the constant.

cl4es replied:
Does the changes in 9800a99 resolve your concerns?

simonis replied:
In principle yes, but shouldn't the condition read:
```cpp
if (!Matcher::match_rule_supported(Op_EncodeISOArray) || !Matcher::supports_encode_ascii_array) return false;
```
I.e. the intrinisc is supported if both conditions are true and not supported if either one of them is false?

cl4es replied:
Resolved in 47633cb

cl4es commented:
> Should we remove the "iso" part from the method/class names?

I'm open to suggestions, but I've not been able to think of anything better. `encodeISOOrASCII` doesn't seem helpful and since ASCII is a subset of the ISO-8859-1 encoding referred to by the "iso" moniker then the ASCII-only variant is technically encoding chars to valid ISO-8859-1.

TobiHartmann commented:
> > Should we remove the "iso" part from the method/class names?
> 
> I'm open to suggestions, but I've not been able to think of anything better. `encodeISOOrASCII` doesn't seem helpful and since ASCII is a subset of the ISO-8859-1 encoding referred to by the "iso" moniker then the ASCII-only variant is technically encoding chars to valid ISO-8859-1.

Okay, that's fine with me.

cl4es commented:
Thanks for reviewing, @TobiHartmann 

I also cleaned up the test and made sure it fails when there's a logic bug like the one I introduced in 9800a99 where the ISO array intrinsic would be matched for a case requiring the ASCII-only array intrinsic. The test was (in)conveniently never testing out-of-range characters in the 0x80-0xFF range, which is precisely where the two intrinsics would produce different results. I hope this doesn't require another 24 hour grace period. 😄

TobiHartmann reviewed (approved):
The incremental change looks good and trivial to me.

cl4es commented:
/integrate
